plot_num_threads_per_disk.py

```diff
@@ -0,0 +1,183 @@
+
+
+def __get_jcts_from_logs_for_phase(event_log_filepath, warmup_count, phase):
+  """
+  Returns a list of JCTs (in seconds) for the specified phase parsed from the specified event log.
+  warmup_count indicates the number of jobs that should be discarded. Assumes that the event log
+  corresponds to one run of org.apache.spark.examples.monotasks.disk.DiskThroughputExperiment.
+  """
+  assert phase in ["write", "read"]
+  filterer = __write_job_filterer if phase == "write" else __read_job_filterer
+  filterer_partial = functools.partial(filterer, warmup_count)
+  return [job.runtime() / 1000.0 for job in
+    parse_event_logs.Analyzer(event_log_filepath, filterer_partial).jobs.itervalues()]
+
+
+def __get_num_threads_from_log_dir(log_dir):
+  """
+  Extracts the number of threads per disk from the specified log directory name. log_dir should
+  be of the form: experiment_log_<experiment parameters>_<num threads per disk>_<timestamp>
+  """
+  return int(log_dir.split("_")[-2])
+
+
+def __create_num_threads_vs_jct_graph(num_threads_to_jcts, output_dir, phase):
+  """
+  Create a graph of num threads per disk vs. JCT for the specified phase, which must be either
+  "write" or "read". num_threads_to_jcts should be a dictionary of the form:
+    { num threads : ( list of write JCTs, list of read JCTs ) }
+  """
+  assert phase in ["write", "read"]
+
+  num_ticks = len(num_threads_to_jcts) + 2
+  xmax = num_ticks - 1
+  max_jct = max([jct
+    for write_jcts, read_jcts in num_threads_to_jcts.itervalues()
+    for jct in (write_jcts if phase == "write" else read_jcts)])
+  ymax = max_jct * 1.1
+  pyplot.title("Num threads per disk vs. JCT (%s phase)" % phase)
+  pyplot.xlabel("Num threads per disk")
+  pyplot.ylabel("JCT (s)")
+  pyplot.grid(b=True)
+  pyplot.xlim(xmin=0, xmax=xmax)
+  pyplot.ylim(ymin=0, ymax=ymax)
+
+  # Build a list of lists of JCTs, sorted by num threads per disk.
+  all_jcts = [write_jcts if phase == "write" else read_jcts
+    for _, (write_jcts, read_jcts) in sorted(num_threads_to_jcts.iteritems())]
+  pyplot.boxplot(all_jcts, whis=[0, 100])
+
+  # Replace the visually-correct x-axis values with the numerically correct values.
+  pyplot.xticks(xrange(num_ticks), [""] + sorted(num_threads_to_jcts.keys()) + [""])
+
+  # Save the graph as a PDF.
+  output_filepath = path.join(output_dir, "%s_phase_num_threads_vs_jct.pdf" % phase)
+  with backend_pdf.PdfPages(output_filepath) as pdf:
+    pdf.savefig()
+
+  pyplot.close()
+
+
+def __write_job_filterer(warmup_count, all_jobs_dict):
+  """
+  The first job generates the data. Of the remaining jobs, the first half are writes. wamrup_count
+  specifies the number of jobs that are warmup and should be discarded.
+  """
+  max_write_job_id = math.ceil(len(all_jobs_dict) / 2.0) - 1
+  return {job_id: job
+    for (job_id, job) in sorted(all_jobs_dict.iteritems())
+    if (job_id > warmup_count) and (job_id <= max_write_job_id)}
```

Comment:
After the write jobs (the last of which has ID max_write_job_id) come all of the read jobs.